asciidtype/asciidtype/src/dtype.c

```diff
@@ -226,6 +226,82 @@ static PyMemberDef ASCIIDType_members[] = {
         {NULL},
 };
 
+static int PICKLE_VERSION = 1;
+
+static PyObject *
+asciidtype__reduce__(ASCIIDTypeObject *self)
+{
+    PyObject *ret, *mod, *obj, *state;
+
+    ret = PyTuple_New(3);
+    if (ret == NULL) {
+        return NULL;
+    }
+
+    mod = PyImport_ImportModule("asciidtype");
+    if (mod == NULL) {
+        Py_DECREF(ret);
+        return NULL;
+    }
+
+    obj = PyObject_GetAttrString(mod, "_reconstruct_ASCIIDType");
+    Py_DECREF(mod);
+    if (obj == NULL) {
+        Py_DECREF(ret);
+        return NULL;
+    }
+
+    PyTuple_SET_ITEM(ret, 0, obj);
+
+    PyTuple_SET_ITEM(ret, 1, Py_BuildValue("(l)", self->size));
+
+    state = PyTuple_New(1);
+
+    PyTuple_SET_ITEM(state, 0, PyLong_FromLong(PICKLE_VERSION));
+
+    PyTuple_SET_ITEM(ret, 2, state);
```

Comment:
I think:
```
Py_BuildValue("O(l)", Py_TYPE(self) self->size,self->size)
```
is good enough? presumably we can consider `ASCIIDType(length)` to be stable API?  And you made the type pickable using the copyreg, so no need for the reconstruct function?  Plus, if we _have_ a reconstruction function, then we certainly can assume stable API?

Reply:
Ah, you didn't do the copyreg for the ASCIIDtype, I think.  But maybe that is nicer?  For the StringDType, this seems true even more so?  We can assume `StringDType()` will always give the same thing, so no need for `__setstate__` logic?

Reply:
I'm not using `copyreg` at all in either dtype. It is a good point though that we don't actually need to save the dtype itself at all though, thanks for the hint.

Reply:
Ah, I misread... somehow I thought one of the reconstructs was to make the class work.  So I guess you need the reconstruct helper until we fix the class pickling.

Reply:
That's exactly right. I originally had it as you suggested, but that didn't work because `ASCIIDType` is an instance of `DTypeMeta`, so python checks the type, sees it's in the `copyreg`, and dutifully calls the pickler numpy registers for the type.